pkg/execution/execution.go

```diff
@@ -120,48 +144,62 @@ func (e *Execution) GetTestInstanceStats(ctx context.Context) (goja.Value, error
 		return nil, errors.New("goja runtime is nil in context")
 	}
 
-	stats := map[string]interface{}{
-		"duration": func() goja.Value {
-			return rt.ToValue(float64(es.GetCurrentTestRunDuration()) / float64(time.Millisecond))
+	ti := map[string]func() interface{}{
+		"duration": func() interface{} {
+			return float64(es.GetCurrentTestRunDuration()) / float64(time.Millisecond)
 		},
-		"iterationsCompleted": func() goja.Value {
-			return rt.ToValue(es.GetFullIterationCount())
+		"iterationsCompleted": func() interface{} {
+			return es.GetFullIterationCount()
 		},
-		"iterationsInterrupted": func() goja.Value {
-			return rt.ToValue(es.GetPartialIterationCount())
+		"iterationsInterrupted": func() interface{} {
+			return es.GetPartialIterationCount()
 		},
-		"vusActive": func() goja.Value {
-			return rt.ToValue(es.GetCurrentlyActiveVUsCount())
+		"vusActive": func() interface{} {
+			return es.GetCurrentlyActiveVUsCount()
 		},
-		"vusMax": func() goja.Value {
-			return rt.ToValue(es.GetInitializedVUsCount())
+		"vusMax": func() interface{} {
+			return es.GetInitializedVUsCount()
 		},
 	}
 
-	obj, err := newLazyJSObject(rt, stats)
-	if err != nil {
-		return nil, err
+	return newInfoObj(rt, ti), nil
+}
+
+// newVUInfo returns a goja.Object with property accessors to retrieve
+// information about the currently executing VU.
+func (mi *ModuleInstance) newVUInfo() (*goja.Object, error) {
+	ctx := mi.GetContext()
+	vuState := lib.GetState(ctx)
+	if vuState == nil {
+		return nil, errors.New("getting VU information in the init context is not supported")
+	}
+
+	rt := common.GetRuntime(ctx)
+	if rt == nil {
+		return nil, errors.New("goja runtime is nil in context")
 	}
 
-	return obj, nil
+	vi := map[string]func() interface{}{
+		"id":        func() interface{} { return vuState.VUID },
+		"idGlobal":  func() interface{} { return vuState.VUIDGlobal },
+		"iteration": func() interface{} { return vuState.Iteration },
+		"iterationScenario": func() interface{} {
+			return vuState.GetScenarioVUIter()
+		},
+	}
+
+	return newInfoObj(rt, vi), nil
 }
 
-func newLazyJSObject(rt *goja.Runtime, data map[string]interface{}) (goja.Value, error) {
-	obj := rt.NewObject()
+func newInfoObj(rt *goja.Runtime, props map[string]func() interface{}) *goja.Object {
+	o := rt.NewObject()
 
-	for k, v := range data {
-		if val, ok := v.(func() goja.Value); ok {
-			if err := obj.DefineAccessorProperty(k, rt.ToValue(val),
-				nil, goja.FLAG_FALSE, goja.FLAG_TRUE); err != nil {
-				return nil, err
-			}
-		} else {
-			if err := obj.DefineDataProperty(k, rt.ToValue(v),
-				goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE); err != nil {
-				return nil, err
-			}
+	for p, get := range props {
+		err := o.DefineAccessorProperty(p, rt.ToValue(get), nil, goja.FLAG_FALSE, goja.FLAG_TRUE)
+		if err != nil {
+			common.Throw(rt, err)
```

Comment:
nit: I would've made this return error and just return it in all the places, which will end up doing the same in the end.

Reply:
Sure, see 264ba62f.